File: lib/src/main/java/vserver/Tool.java
package vserver;

import vjson.JSON;
import vjson.ex.JsonParseException;
import vjson.util.ObjectBuilder;
import vproxybase.util.ByteArray;
import vserver.util.UTF8ByteArrayCharStream;

public class Tool {
    public static final RoutingContext.StorageKey<JSON.Instance> bodyJson = new RoutingContext.StorageKey<>() {
    };

    public static RoutingHandler bodyJsonHandler() {
        return ctx -> {
            ByteArray body = ctx.body();
            if (body != null) {
                JSON.Instance inst;
                try {
                    inst = JSON.parse(new UTF8ByteArrayCharStream(body));
                } catch (JsonParseException e) {
                    ctx.response().status(400)
                        .end(new ObjectBuilder()
                            .put("status", 400)
                            .put("reason", "Bad Request")
                            .put("message", "request body is not valid json: " + e.getMessage())
                            .build());
                    return;
                }
                ctx.put(bodyJson, inst);
            }
            ctx.next();
        };
    }

    private Tool() {
    }
}

File: lib/src/main/kotlin/io/vproxy/lib/http/Tool.kt
package vproxy.lib.http

import vjson.JSON
import vjson.JSON.parse
import vjson.cs.UTF8ByteArrayCharStream
import vjson.ex.JsonParseException
import vjson.util.ObjectBuilder

object Tool {
  @JvmStatic
  val bodyJson: StorageKey<JSON.Instance<*>> = object : StorageKey<JSON.Instance<*>> {}
  private val bodyJsonHandlerInstance = object : RoutingHandler {
    override suspend fun handle(rctx: RoutingContext) = handleBodyJson(rctx)
  }

  @JvmStatic
  fun bodyJsonHandler(): RoutingHandler {
    return bodyJsonHandlerInstance
  }

  private suspend fun handleBodyJson(ctx: RoutingContext) {
    val body = ctx.req.body()
    if (body.length() != 0) {
      val inst: JSON.Instance<*>
      try {
        inst = parse(UTF8ByteArrayCharStream(body.toJavaArray()))
      } catch (e: JsonParseException) {
        ctx.conn.response(400).send(
          ObjectBuilder()
            .put("status", 400)
            .put("reason", "Bad Request")
            .put("message", "request body is not valid json: " + e.message)
            .build()
        )
        return
      }
      ctx.put(bodyJson, inst)
    }
    ctx.allowNext()
    return
  }
}
